pandas/tests/frame/test_stack_unstack.py

```diff
@@ -1052,6 +1052,30 @@ def test_stack_preserve_categorical_dtype(self, ordered, labels):
 
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize("ordered", [False, True])
+    @pytest.mark.parametrize("labels", [list("yxz"), list("yzx")])
+    @pytest.mark.parametrize("labels2", [list("uv"), list("vu")])
+    def test_multi_stack_preserve_categorical_dtype(self, ordered, labels, labels2):
+        # GH-36991
+        cidx = pd.CategoricalIndex(labels, categories=list("xyz"), ordered=ordered)
+        cidx2 = pd.CategoricalIndex(labels2, categories=list("uv"), ordered=ordered)
+        sorted_cidx = pd.CategoricalIndex(
+            list("xyz"), categories=list("xyz"), ordered=ordered
+        )
+        sorted_cidx2 = pd.CategoricalIndex(
+            list("uv"), categories=list("uv"), ordered=ordered
+        )
+
+        midx = MultiIndex.from_product([cidx, cidx2, [1, 2, 3]], names=list("abc"))
+        df = DataFrame(np.random.randn(5, midx.size), columns=midx)
+        result = df.stack(["a", "b"])
+
+        expected = MultiIndex.from_product(
+            [df.index, sorted_cidx, sorted_cidx2], names=[None, "a", "b"]
+        )
+
+        tm.assert_equal(result.index, expected)
```

Comment:
can you compare the resulting frame itself

Reply:
Done.